lib/render/ParticleRenderer.cpp

```diff
@@ -178,4 +190,217 @@ int ParticleRenderer::_paintGL(bool)
     return rc;
 }
 
-int ParticleRenderer::_initializeGL() { return 0; }
+int ParticleRenderer::_initializeGL() { 
+    glGenVertexArrays(1, &_vertexArrayId);
+    glGenBuffers(1, &_vertexBufferId);
+
+    /* Generate and configure 1D texture: _colorMapTexId */
+    glGenTextures(1, &_colorMapTexId);
+    glActiveTexture(GL_TEXTURE0 + _colorMapTexOffset);
+    glBindTexture(GL_TEXTURE_1D, _colorMapTexId);
+    glTexParameteri(GL_TEXTURE_1D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+    glTexParameteri(GL_TEXTURE_1D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+    glTexParameteri(GL_TEXTURE_1D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+    glBindTexture(GL_TEXTURE_1D, 0);
+
+    glGenVertexArrays(1, &_VAO);
+    glGenBuffers(1, &_VBO);
+
+    assert(_VAO);
+    assert(_VBO);
+
+    glBindVertexArray(_VAO);
+    glBindBuffer(GL_ARRAY_BUFFER, _VBO);
+    glVertexAttribPointer(0, 3, GL_FLOAT, GL_FALSE, sizeof(glm::vec4), NULL);
+    glVertexAttribPointer(1, 1, GL_FLOAT, GL_FALSE, sizeof(glm::vec4), (void *)sizeof(glm::vec3));
+    glEnableVertexAttribArray(0);
+    glEnableVertexAttribArray(1);
+    glBindBuffer(GL_ARRAY_BUFFER, 0);
+    glBindVertexArray(0);
+    return 0; 
+}
+
+int ParticleRenderer::_renderParticles(const std::vector<glm::vec4>& particles)
+{
+    typedef struct {
+        glm::vec3  p;
+        float v;
+    } Vertex;
+    vector<Vertex> vertices;
+    vector<int>    sizes;
+    vector<Vertex> sv;
+
+    int nStreams = particles.size();
+    for (int s = 0; s < nStreams; s+=2) {
+        glm::vec4 p = particles[s];
+        glm::vec4 p2 = particles[s+1];
+        const vector<flow::Particle> stream = {flow::Particle(p[0],p[1],p[2],0.,p[3]), flow::Particle(p2[0],p2[1],p2[2],0.,p2[3])};
+        sv.clear();
+        int sn = stream.size();
+
+        for (int i = 0; i < sn + 1; i++) {
+            // "IsSpecial" means don't render this sample.
+            if (i == sn || stream[i].IsSpecial()) {
```

Comment:
Why is this code here?

Reply:
Like you mentioned, an idiosyncrasy in the flow renderer.  I removed as much flow code as possible, but this section is hard to understand so I treated it like a black box.

@shaomeng, do you know if there are some more descriptive variable names I can substitute here for "sv", "sn", and "svn"?  Also, can you talk about what "IsSpecial" means?

Reply:
Sure. In flow renderer, a "flow line" is just a sequence of particles. When you render them, you draw a line from one particle to the next. Occasionally, the flow line is disconnect, for example, it exits from one side of the volume and re-enters from the other side. In this case, I put a "special" particle in the sequence to mark that you want to skip that particular particle and restart drawing lines from the next one. Does this make sense?

Reply:
@shaomeng Yes I think so.  Thanks.  Sounds like I just need to remove the "IsSpecial" special case at line 243.